doc/decisions/1_in_discussion/internal_cache.md

```diff
@@ -0,0 +1,350 @@
+  ```c
+  Key * originalKey;
+  Key * copiedKey;
+  keyCopy (copiedKey, originalKey, ELEKTRA_CP_COW);
+  
+  assert (keyString (copiedKey) == keyString (originalKey));
+  assert (keyName (copiedKey) == keyName (originalKey));
+
+  keyDel (originalKey);
+  
+  keyString (copiedKey); // Error! Original value has been deleted. Pointer to data in copiedKey points to freed memory
+  keyName (copiedKey);   // Error! Original name has been deleted.
+  ```
+ 
+  Triggering the update problem:
+  ```c
+  Key * originalKey;
+  Key * copiedKey;
+  keyCopy (copiedKey, originalKey, ELEKTRA_CP_COW);
+  
+  assert (keyString(copiedKey) == keyString(originalKey));
+  keySetString (originalKey, "new value!");
+  
+  keyString(copiedKey); // Error! Original value has been deleted. Pointer to data in copiedKey points to potentially freed memory
+  ```
+
+  The same problems in principle exist for `mmapstorage`, however this is just a storage plugin and the user does not have any API to create COW keys themselves.
+  We can still let users access the flag `ELEKTRA_CP_COW`, we just need to clearly document what is forbidden.
+  Maybe set the `KEY_FLAG_RO_VALUE` on the original key, so that the API itself detects the error. 
+  There is, however, no flag for `keyDel` that we could set.
+
+### Data restrictions
+
+@kodebach wrote:
+
+> Make all the keys returned by kdbGet completely read-only.
+> To change the data you need to append an entirely new key to replace the existing one.
+> Then we just need to keep a shallow copy internally.
+
+### API restrictions
+
+@kodebach wrote:
+
+> Change the API and remove KeySet from kdbGet and kdbSet also option 4 in [the operation sequences decision](../0_drafts/operation_sequences.md).
+> If the keyset is owned by the KDB handle, it should not be as big surprise, if there is extra data in there.
+> I certainly wouldn't try to asset anything on the contents of a KeySet that I don't own directly, unless the condition is explicitly documented somewhere.
+
+@markus2330 wrote:
+
+> I disagree, it is actually the same kind of surprise for "More Keys".
+> Only the "Fewer Keys" would get fixed.
+
+### Full-blown copy-on-write implementation
+
+Make Elektra's `Key` and `KeySet` data structures copy-on-write.
+This requires some major refactoring of code within `libelektra-core`.
+Code that does only interact with the data structures via the public `libelektra-core` API should not notice any differences.
+The `mmapstorage` plugin will need a major refactoring.
+
+#### Changes to `Key`
+
+For the `Key`, we need to extract everything for the data and name into their own structs.
+This is done for memory-management reasons, as we need to track how many keys point to the same data and/or name.
+
+```c
+struct _KeyData {
+    union {
+        char * c;
+        void * v;
+    } data;
+
+    size_t dataSize;
+
+    uint16_t refs;
+};
+
+struct _KeyName {
+    char * key;
+    size_t keySize;
+
+    char * ukey;
```

Comment:
I have an additional proposal here: What about having an extra namespace, several (ref-counted) pointers to namespace, dirname and basename. The dirname is what we have as parent key. Then we need to store the parent-key (usually the largest portion) only once, and for all the other keys only the basename.

This way should save plenty of memory (instead of using more). The problem is that we cannot have a `keyName` anymore.

Reply:
Storing the namespace separately could make sense (for different reasons), but that's just a single byte.

Regarding `dirname` & `basename`:
There are many problems with this. The first and biggest is where do we split between `dirname` and `basename`? Finding the optimal split would be quite difficult and would change over time. How would the API for key creation and key name changes look like?

If we really want to go in that direction, I think we actually need to rethink the whole API (again) and see, if we can find a better way to do things.

Reply:
Nice proposal, but let's try to avoid doing everything at once in this decision.

Reply:
> It is not a single byte. With COW is N*M bytes, N the number of namespaces, M the shared portion of the keyname.

Assuming we remove the escaped key name, the namespace is stored literally as a single byte. There is no point in doing anything other than just copying it.

> I suggested the parent-key above.

But what is _the parent-key_? The term is context-dependent. There is the parent key of a mountpoint, the parent key used in `kdbGet`, etc.

Even if one of those is chosen (parent of mountpoint would be sensible), this still doesn't answer how `keyNew` or `keySetName` would now about this.

> I don't see a connection between "not having keyName (and an iterator instead)" and "need to rethink the whole API".

And that's honestly, why you shouldn't make these decisions. I've dealt with key name changes before. I know it affects _literally everything_. If there is any point to rethink the API that is the point. It is almost impossible to find an API change that would affect more code than key name changes.

And like I explained above, but you may not have realized yet, the changes you proposed **require** API changes. Not just that, but API changes to pretty fundamental functions used all over the code base. This really would be the point to rethink the whole API.

> mpranj can you add these possibilities to the store_name.md decision?

I seriously doubt mpranj wants to be involved in this. AFAIK the task isn't even related to mpranj's previous work. IMO it is a very bad habit of yours to assign such tasks to random people.

Reply:
Since you marked my comment as disruptive, I'll repeat the purely technical and absolutely correct parts again. There is no need to answer it here and now. But if you expect this approach to be added to the decision or to be discussed elsewhere, we need an answer to these questions.

> It is not a single byte. With COW is N*M bytes, N the number of namespaces, M the shared portion of the keyname.

Assuming we remove the escaped key name, the namespace is stored literally as a single byte. There is no point in doing anything other than just copying it.

> I suggested the parent-key above.

But what is _the parent-key_ here? The term is context-dependent. There is the parent key of a mountpoint, the parent key used in kdbGet, etc.

Even if one of those is chosen (parent of mountpoint would be sensible), this still doesn't answer how keyNew or keySetName would now about this. I don't see how this would work without API changes to very basic APIs used all over the code base.